include/zephyr/shell/shell.h

```diff
@@ -1241,6 +1290,160 @@ int shell_set_root_cmd(const char *cmd);
+ * @param[in]  data		Address to place the received data.
+ * @param[out] bytes_to_read	Maximum number of data bytes to receive.
+ * @param[in]  bytes_read	Address of area to hold the number of bytes
+ *				received.
+ * @param[in]  min_xfer		Minimum number of data bytes to receive.
+ * @param[in]  timeout		Waiting period to wait for the data to be
+ *				received, or one of the special values K_NO_WAIT
+ *				and K_FOREVER.
+ *
+ * @retval 0			At least @a min_xfer bytes of data were
+ *				received.
+ * @retval -EPERM		Invoked from invalid thread.
+ * @retval -EINVAL		Invalid parameters supplied.
+ * @retval -EIO			Returned without waiting; zero data bytes were
+ *				received.
+ * @retval -EAGAIN		Waiting period timed out; between zero and
+ *				@a min_xfer minus one data bytes were received.
+ */
+int shell_bypass_receive(struct shell_bypass *shb, void *data,
+			 size_t bytes_to_receive, size_t *bytes_received,
+			 size_t min_xfer, k_timeout_t timeout);
+
+/**
+ * @brief Send bytes to the shell's underlying transport layer.
+ *
+ * The caller must obtain a valid struct shell_bypass by entering user bypass
+ * mode with shell_bypass_enter(). Only the thread that invoked the enter
+ * method may call shell_bypass_send().
+ *
+ * This function uses a pipe whose size and alignment are specified by compile-
+ * time configuration variables SHELL_BYPASS_USER_TX_PIPE_SIZE and
+ * SHELL_BYPASS_USER_TX_PIPE_ALIGN, respectively. A pipe buffer whose size is
+ * too small may result in increased CPU overhead on large data sends due to
+ * frequent thread context switching.
+ *
+ * @param[in]  shb		Pointer to shell user bypass structure.
+ * @param[in]  data		Address of data to send
+ * @param[in]  bytes_to_send	Size of data (in bytes).
+ * @param[out] bytes_sent	Address of area to hold the number of bytes
+ *				sent.
+ * @param[in]  min_xfer		Minimum number of bytes to write.
+ * @param[in]  timeout		Waiting period to wait for the data to be
+ *				written, or one of the special values K_NO_WAIT
+ *				and K_FOREVER.
+ *
+ * @retval 0			At least min_xfer bytes of data were sent.
+ * @retval -EPERM		Invoked from invalid thread.
+ * @retval -EIO			Returned without waiting; zero data bytes were
+ *				sent.
+ * @retval -EAGAIN		Waiting period timed out; between zero and
+ *				min_xfer minus one data bytes were sent.
+ */
+int shell_bypass_send(struct shell_bypass *shb, const void *data,
+		      size_t bytes_to_send, size_t *bytes_sent,
+		      size_t min_xfer, k_timeout_t timeout);
+
+/**
+ * @brief Exit user bypass mode.
+ *
+ * The caller must obtain a valid struct shell_bypass by entering user bypass
+ * mode with shell_bypass_enter(). Only the thread that invoked the enter
+ * method may call shell_bypass_exit().
+ *
+ * Any data sent via shell_bypass_send() that has not yet been flushed to the
+ * underlying transport will be dropped. Any data received by the transport
+ * layer but not yet received with shell_bypass_receive() will be dropped.
```

Comment:
seems like a flush (block until output finished) function or function for getting remaining data might be useful